src/monitor/node_active_protocol.c

```diff
@@ -1137,76 +1135,109 @@ start_maintenance(PG_FUNCTION_ARGS)
 	if (currentNode->reportedState == REPLICATION_STATE_MAINTENANCE ||
 		currentNode->goalState == REPLICATION_STATE_MAINTENANCE)
 	{
-		ereport(ERROR,
-				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
-				 errmsg("cannot start maintenance: "
-						"node %s:%d is already in maintenance",
-						currentNode->nodeName, currentNode->nodePort)));
+		/* if we're already in maintenance, we're good */
+		PG_RETURN_BOOL(true);
 	}
 
-	if (!(IsStateIn(currentNode->reportedState, secondaryStates) &&
-		  currentNode->reportedState == currentNode->goalState))
+	/*
+	 * We allow to go to maintenance in the following cases only:
+	 *
+	 *  - current node is a primary, and we then promote the secondary
+	 *  - current node is a secondary or is catching up
+	 */
+	if (!(IsCurrentState(currentNode, REPLICATION_STATE_PRIMARY) ||
+		  (IsStateIn(currentNode->reportedState, secondaryStates))))
 	{
 		ereport(ERROR,
 				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
 				 errmsg("cannot start maintenance: current state for node %s:%d "
-						"is \"%s\", expected either "
+						"is \"%s\", expected either \"primary\", "
 						"\"secondary\" or \"catchingup\"",
 						currentNode->nodeName, currentNode->nodePort,
 						ReplicationStateGetName(currentNode->reportedState))));
 	}
 
-	if (!(IsStateIn(otherNode->goalState, primaryStates) &&
-		  currentNode->reportedState == currentNode->goalState))
+	if (IsCurrentState(currentNode, REPLICATION_STATE_PRIMARY))
+	{
+		/*
+		 * We put the primary to maintenance now, and when it has reached the
+		 * assigned goal then the group state machine in ProceedGroupState() is
+		 * going to drive the secondary straight to WAIT_PRIMARY.
+		 */
+		LogAndNotifyMessage(
+			message, BUFSIZE,
+			"Setting goal state of %s:%d to maintenance "
+			"after a user-initiated start_maintenance call.",
+			currentNode->nodeName, currentNode->nodePort);
+
+		SetNodeGoalState(currentNode->nodeName, currentNode->nodePort,
+						 REPLICATION_STATE_MAINTENANCE);
+
+		NotifyStateChange(currentNode->reportedState,
```

Comment:
Unrelated (so probably better in separate PR), but could this function be refactored to take an `AutoFailoverNode`? Instead of all the fields of it separately. Currently it is hard to review that the right arguments have been passed in.

Or even better, call this from `SetNodeGoalState` (which should then take an `AutoFailoverNode`). So calling it cannot be forgotten.

Reply:
Good point, and agreed: let's see about that in another PR.

Reply:
Maybe we should also switch the notification format (and parser) to using JSON by the way. We now have a JSON parser in our code base.